### ISSUE TITLE
Securing the Kata Control Plane

### ISSUE THREAD

**Issue by ray-valdez** [confidential-containers/community#53]
### Problem
 
Confidential containers use kata to deploy and run containers in confidential VMs. This approach protects container workloads against privileged software and malicious administrators on the host. Presently, the kata control plane, which consists of containerd and kata-runtime, is vulnerable to attacks. Kata-runtime sends commands to the kata-agent inside confidential VMs. These commands are exposed to untrustworthy hosting platforms.  A malicious CSP or administrator can intercept/compromise commands from kata control plane, tamper with the container config file, execute processes on the container, retrieve workload stats, or obtain sensitive container workload log.

### Proposed Solution
 
To mitigate the attack on the kata control plane, the confidential container team has proposed to restrict the kata-agent API endpoints by disabling sensitive commands, such as ExecProcess, ReseedRandomDev, and telemetry. Similarly, our objective is to minimize the attack surface of the kata control plane and only permit CSPs to launch or terminate sandbox VMs.  This is comparable to the threat model of confidential VMs. 
 
We propose to remove basically all commands from the untrustworthy kata control plane, except for ones starting and terminating the sandbox VM, and introduce another secure channel for accessing the kata-agent APIs. This secure channel can be set up in a similar fashion as the communication between attestation-agent and key broker service. We can establish a secure TLS channel after successfully attesting the integrity of confidential VMs and only allow authenticated container owner to deliver commands to kata-agent via this channel.  Therefore, we can ensure the confidentiality and integrity of commands in the flight. Only the container owner can directly manage and monitor container workloads in confidential VMs.

**Comment by bpradipt** [comment]
@ray-valdez thanks for creating this issue. This is an important aspect. A quick question. Where do you envisage the control plane that is interacting with the kata-agent to be running ? 

@c3d I'm wondering if this is on similar lines to a separate control plane  (`kcp` based ??) that you had discussed in one of your earlier presentations.

**Comment by c3d** [comment]
@bpradipt @ray-valdez Yes, at least from the description, this sounds like a first step towards what is described in https://docs.google.com/presentation/d/1TDFwrU3rdkXSTw0XYlmuV1VzSC4GkdI2bteq5GDmgAo/edit,

**Comment by c3d** [comment]
@ray-valdez Please feel free to borrow any of the slides to illustrate your proposal.

**Comment by ray-valdez** [comment]
> @ray-valdez thanks for creating this issue. This is an important aspect. A quick question. Where do you envisage the control plane that is interacting with the kata-agent to be running ?
> 
> @c3d I'm wondering if this is on similar lines to a separate control plane (`kcp` based ??) that you had discussed in one of your earlier presentations.

**Comment by ray-valdez** [comment]
We envision running dual control planes. One control plane is provided by the CSP for managing VM level functionality/life-cycle, e.g., starting the sandbox VM and the other by the tenant that will interact with the kata-agent. I agree this similar to @c3d's tenant splitting APIs.

**Comment by ray-valdez** [comment]
> @ray-valdez Please feel free to borrow any of the slides to illustrate your proposal.

@c3d Thanks for providing the presentation. What is the status of your tenant split APIs work? We share the same objectives, and thus there are possibilities for collaboration.

**Comment by gzs715** [comment]
@bpradipt One potential solution is to partition the endpoint APIs of kata-agent into two parts. We still allow the CSP to launch/terminate a VM. This is compatible with the confidential VM's threat model. For the other endpoints, it should be exposed only to the authenticated tenant after remote attestation. We are thinking about reusing the rats-tls session in the attestaion-agent. Once the key is provisioned, the tenant can continue to control the workloads within the sandbox via the rats-tls channel.

**Comment by c3d** [comment]
@ray-valdez This was only at the planning stage at this point. We had several iterations of CC (v0, v1, etc), and this was not in the v1 time frame, so no actual code or anything. Some analysis of the APIs that needed to be changed and how. Part of the problem is some existing OCI APIs having fields that are "in the wrong place", and those are hard to change. An example is the I/O descriptors (stdin, stdout, stderr)

**Comment by ray-valdez** [comment]
**Establishing a Secure Channel**
As a first step for splitting the host and tenant APIs, we need to provide a secure channel for the tenant, i.e., confidential container owner, to send commands to the kata-agent for managing the container life cycle inside of the confidential VM.
 
Since the KBS (relying party) is trusted and establishes an attested TLS secure channel with the attestation agent in the confidential VM, we can use its `get_resource API ` to provision secrets onto the confidential VM for setting up a secure channel with the tenant management stack.  The secrets can consist of a private/public key pair for the server running inside of the confidential container and a client credential for a remote client running on the tenant stack. The client credential is used to verify the client’s public key.
 
After the attestation agent establishes a secure channel with the KBS, the kata-agent requests channel secrets via KBS’s `get_resource API`, see step 1 in the figure.  Then the kata-agent spawns a gRPC API proxy server configured with the server’s TLS credential and client credential. The tenant then establishes a TLS channel with the gRPC API proxy in which they mutually authenticate each other, see step 2 in the figure. The gRPC API proxy serves to route commands from the tenant to the kata-agent. 
 
 
A design Issue
 
Although this design avoids baking secrets into VM images for tenants, we still need to distinguish VMs from different tenants.  In the cloud use case, the CSP starts generic confidential VMs (provisioned using the same kernel, root file system and kata agent), this means that different tenants will have the same sandbox VM measurements if they share the same key broker. This is not an issue if each tenant has its own distinct key broker.
 
In the use case where we share a key broker, we could depend on the CSP’s assignments of MAC and IP to tenant to distinguish VMs from different tenants. This association can be leveraged by the KBS to provision the VM with correct channel secrets. 
![establishing_secure_channel](https://user-images.githubusercontent.com/105830165/176336327-adab9a0e-53e2-43aa-959d-34520f7f5888.png)

**Comment by c3d** [comment]
@ray-valdez A few comments on the diagram above, notably based on the discussion we had in the meeting on January 27, 2022.

1. We discussed two options for receiving commands through a secure channel, one is to extend the attestation channel so that it can send commands, the other being to use the attestation's ability to deliver secrets in order to setup a separate channel. You are describing the second approach in the above. I think we agreed this was probably a simpler design.
2. Regarding the TLS secrets being delivered, they could be ephemeral, i.e. the key broker should generate a new private/public key pair for every new attestation. This will notably minimize the exposure to replay attacks.
3. You suggest a differentiation based on MAC addresses, but to the extent that the TLS client uses an IP and not a MAC address, that is probably the association the tenant stack should keep.
4. In your diagram, you seem to dissociate "relying party" and "tenant stack". In my mind, the tenant stack was a part of the relying party, much like the key broker. As a matter of fact, because of the IP/TLS keys association above, I would argue that the tenant stack needs write access to the key broker.

**Comment by ray-valdez** [comment]
@c3d a reply to your comments:
> @ray-valdez A few comments on the diagram above, notably based on the discussion we had in the meeting on January 27, 2022.

2. I agree the use of an ephemeral key pair strengthens the security against replay attack.
 
3. Yes, TLS uses IP addresses, and the MAC address cannot be used outside of a local system to differentiate tenant VMs. My point here is that CSP assigns some key  attribute, e.g., IP, that we must depend on. 
 
4. My initial thought was that the relying party could provide trusted services to multiple different tenants. As you highlight in your first point for IP/TLS key association, there must be some close cooperation between the relying party and tenant stack for setting up channel parameters.  You make a reasonable point. Incorporating the "tenant stack" as part of the “relying party" will simplify the design.

**Comment by gzs715** [comment]
@ray-valdez @c3d For point 3:
MAC address is at the device level. Considering that live migration will be supported in the future, MAC address may change. IP would be a better attribute for a VM.

**Comment by ray-valdez** [comment]
Due to the timing order of commands and pod/containers state info maintained on the host side, it difficult to remove completely the vsock channel without re-working dispatching of commands from the side host to the kata-agent.  As a first step, we focus on prototyping sensitive commands (e.g., exec) over a secure channel to gain more understanding of the issues.

**This prototype work involves the following:** 
1. Embed a gRPC API proxy (tls server) into the kata-agent which forwards requests to the appropriate sandbox functions;
2. Modify agent-ctl to forward commands to the gRPC API proxy; and
3. Employ KBS’s `get_resource` API to provision secrets (server private/public keys and client credential) to the gRPC API proxy.

**Identified todos:** 
- Understand mapping of shim v2 endpoints to kata-agent endpoints
- Investigate alternatives for implementing the tenant side of the stack
`-` oci plex (forked) since it can dispatch to different backends:  oci command line;  shim v2; and kata-agent
`-` runk, it’s a runtime implementing the kata-agent code and may already have abstracted the interface for talking to the kata-agent

**Comment by ray-valdez** [comment]
Summary of 08-08-22 discussions

Identified three options for providing the tenant-side of the management stack:
1.    a (modified) version of `runk`, see https://github.com/kata-containers/kata-containers/tree/main/src/tools/runk;
2.   `oci plex` (based on fork of https://github.com/containers/youki) to dispatch commands to different backends (either the shim v2 runtime on the host or the secure channel to the kata-agent) @c3d could you a provide link to `oci plex`?
3.  Dispatching commands at a high level using either `kubectl `or `oc` with different sets of credentials for host and tenant management stacks. Security sensitive commands, such as `ExecProcess` or logs, would use the tenant side and other commands, such as pod create would use the host side.
Note all options will require commands to be sent over the secure channel to the kata-agent at some point.

Started experimenting using tonic to implement  gRPC TLS server and client for confidential containers. 

- Forked kata containers and created a feature branch `split-api-feature`  from the CCv0 branch, see https://github.com/ray-valdez/kata-containers
-  As a step first, we embedded secrets into the kata-agent and implemented the `PauseContainerRequest` (since it only requires a single argument, container id), testing connectivity with the kata-agent using `kata-containers/src/tools/cmd_tls_ctl`
- Evaluated running `ExecProcess` using  `kata-containers/src/tools/agent-ctl`.  Note this command must not use any host based IO when requested over the secure channel. 

Identified the need to separate log messages from VM, kata-agent and container(s). Currently kata-containers multiplexes different log messages onto the same file descriptor.

**Comment by ray-valdez** [comment]
Two major points were discussed in our last meeting on 08-22-22.

- Both peer pods and host/tenant split APIs address a similar problem. However, they partition the host and tenant control planes slightly differently.  We want to leverage the peer pods work since launching a confidential VM can be viewed as a special case of peer pods.  We also want the same solution for the dispatching problem, i.e., sending tenant commands to the confidential VM, for both.  @c3d will sync with @bpradipt and provide a follow up on leveraging peer pods.

-  In the host/tenant split APIs, there is a point (in time) when control is transferred to the tenant from the host.  The tenant then needs to acquire information on what’s deployed (i.e., sandbox) by querying the kata-agent. Ideally, the transfer can occur when the tenant establishes a secure channel with the  kata-agent.

**Comment by ariel-adam** [comment]
@ray-valdez is this issue still relevant or can be closed? 
If it's still relevant to what release do you think we should map it to (mid-November, end-December, mid-February etc...)?

**Comment by ray-valdez** [comment]
@ariel-adam I think the issue is relevant; it's not solved. We found a work-around but containerd's CC-main branch has a problem for confidential containers, see [issue #59](https://github.com/confidential-containers/community/issues/59). I am not sure about the targeted release because it depends on plans to support Ubuntu 22.04 as the host OS.

**Comment by ariel-adam** [comment]
Got it.
I've mapped it to the CoCo releases project with no drop so it's part of the backlog.
If we see it still in the backlog in 9-12 month lets consider closing it.

**Comment by c3d** [comment]
Updated slide deck from the discussion last week
https://docs.google.com/presentation/d/16649JpQAdDb3jh3OVKWdNks0mvcqpidssrm0pOZNKZ4/edit?usp=sharing

**Comment by fitzthum** [comment]
btw one other aspect of this that we haven't talked about much is way that the shared filesystem is used. This is how`resolv.conf` is passed into the guest, for instance. Generally speaking the API of the guest is more than just the API of the kata agent.

**Comment by ray-valdez** [comment]
@fitzthum Why is this so? The network plugin,`cni`,  already provides the network configuration: gateway, subnet, IP address and so on. It only supports dns configuration. Why not configure the kata-agent/sandbox with the dns configuration provided by the network plugin?

> btw one other aspect of this that we haven't talked about much is way that the shared filesystem is used. This is how`resolv.conf` is passed into the guest, for instance. Generally speaking the API of the guest is more than just the API of the kata agent.

**Comment by fitzthum** [comment]
@ray-valdez I think Kata does use the network configuration provided by CNI. Kata shared is what is used to send this to the guest. CNI is untrusted. Is there any `resolv.conf` that a malicious host could provide to the user that would result in leaking secrets?

**Comment by gzs715** [comment]
@fitzthum Can we just consider all I/O untrustworthy in confidential computing? We can let the guests validate the untrusted inputs, e.g., files in shared storage and network packets from CNI, before using them. Reading may not leak secrets, but guests should be careful about writing/executing files in untrusted shared storage.

**Comment by fitzthum** [comment]
Sure, but how does the guest validate all these inputs and what even is the full scope of the I/O?

**Comment by gzs715** [comment]
For using these inputs from the untrusted shared storage, the guest needs to have a defined security policy or known hash values to validate the files. Or we can let the guest retrieve encrypted files from shared storage. After remote attestation, the guest can get the key from KBS to verify/decrypt them. 
For the full scope of I/O, confidential containers need to rely on end-to-end crypto, e.g., TLS or an encrypted file system, to protect I/O at runtime.

**Comment by fitzthum** [comment]
This gets pretty tricky with some of the more dynamic values such as network configuration, which is generally meant to mirror the host. `resolv.conf` isn't too complicated but even that seems difficult for a guest owner to validate or know ahead of time.

**Comment by gzs715** [comment]
The operations of reading configurations themselves may not leak the secrets if we always keep in mind that the configurations are not trustworthy. We can rely on crypto mechanisms to seal the data before putting them on untrusted mediums. Guest may not control if the nameserver is malicious in the resolv.conf. Even if it is a legitimate nameserver, the DNS request can also be hijacked. But the guest can control that all data exposed to untrusted mediums should have been encrypted, e.g., using DNSSEC. Leaking encrypted data will not break the confidentiality guarantees.

**Comment by fitzthum** [comment]
I generally agree that DNS isn't the most grave case. It seems like maybe manipulating `resolv.conf`can't lead to secrets being leaked, but I am no confident about that. Like I said in the meeting a few weeks ago I think `UpdateRoutes` is that pathological case. That's a dynamic configuration and I don't see how it can be supplied or validated by the guest owner.

**Comment by ray-valdez** [comment]
@fitzthum I don't think that manipulating `resolv.conf` can lead to secrets being leaked. This file basically configures the local DNS resolver. In the worst case scenario, misconfiguration of the network can lead to a denial-of-service attack. As long as communication is over a secure channel, such as TLS designed for insecure network environments, we can prevent information leakage. With respect to `resolv.conf`, I think we need integrity protection more than confidentiality.

**Comment by fitzthum** [comment]
@ray-valdez I generally agree regarding DNS, but I'm not totally confident about all the different options available in `resolv.conf`. I brought it up here mainly as an example of part of the attack surface that is not the Kata Agent API. I am also generally worried about some of the dynamic network config things, but like I said DNS should be safe in theory.

**Comment by c3d** [comment]
@fitzthum Sorry for joining the discussion late, I'm currently fighting some kind of virus…

I think you raise a very good point, namely that we generally need to consider any file passed by the host as potentially hostile, and carefully consider the attacks each of the file enables. While at first sight, I would agree with @ray-valdez that `resolv.conf` is not easily usable to leak secret, we should not forget that hackers tend to use the tools at their disposal in a non-conventional, unintended ways. So I would certainly not rule out that you can devise an interesting attack just based on the network configuration mechanisms.

The domain search list, in particular, or the `NOTFOUND=return` option in `nsswitch.conf`, for example, could easily lead to leaking information about the servers you use. I see that as the kind of data leak we would like to avoid. But then, AFAIK, we do not pass that from the host today.

So I would rephrase the problem as follows: how can we setup the guest networking so that it's compatible with what the host needs, but still lets the guest owner control DNS lookup.

BTW, do we have a list of the files that are bind-mounted from the host? A quick grep through the code did not help me, because we have multiple helpers to mount.

**Comment by fitzthum** [comment]
> BTW, do we have a list of the files that are bind-mounted from the host? A quick grep through the code did not help me, because we have multiple helpers to mount.

I am not aware of any such list. You might try starting a pod with Kata and seeing what is in the shared directory. Ideally we will develop a comprehensive view of the attack surface of the guest, but it's a bit tricky to do that.

**Comment by ray-valdez** [comment]
We have implemented the following kata-agent endpoint APIs to support pod/container creation from the tenant-side, please see issue [#1834](https://github.com/kata-containers/kata-containers/issues/1834) where we categorize and discuss the partitioning of the endpoint APIs between the host-side and tenant-side:
-  PullImage
-  CreateContainer
-  ExecProcess 
-  PauseContainer 
-  RemoveContainer
-  ResumeContainer
-  SignalProcess
-  StartContainer
- WaitProcess
-  ListContainers (new API added to support split host/tenant APIs)
